torch/csrc/api/include/torch/nn/functional/loss.h

```diff
@@ -26,6 +26,36 @@ inline Tensor cosine_embedding_loss(
       input1, input2, target, options.margin(), options.reduction());
 }
 
+inline Tensor _smooth_l1_loss(const Tensor& input, const Tensor& target) {
+    auto t = torch::abs(input - target);
+    return torch::where(t < 1, 0.5 * torch::pow(t, 2), t - 0.5);
+}
+
+inline Tensor smooth_l1_loss(
+    const Tensor& input,
+    const Tensor& target,
+    const SmoothL1LossOptions& options) {
+  
+  if (target.sizes() != input.sizes()) {
+    TORCH_WARN("Using a target size (", target.sizes(), ") that is different to the input size (", input.sizes(), "). ",
+                  "This will likely lead to incorrect results due to broadcasting. ",
+                  "Please ensure they have the same size.");
+  }
+
+  Tensor ret;
+
+  if (target.requires_grad()) {
+    ret = _smooth_l1_loss(input, target);
+    if (options.reduction() != Reduction::None) {
+      ret = options.reduction() == Reduction::Mean ? torch::mean(ret) : torch::sum(ret);
+    }
+  } else {
+        std::vector<Tensor> expanded_tensors = torch::broadcast_tensors({input, target});
+        ret = torch::smooth_l1_loss(expanded_tensors[0], expanded_tensors[1], options.reduction());
```

Comment:
nit: it would be awesome to align the indentation with the first branch of the `if` block :D